Create Blog “scale-kubernetes-clusters-using-hpe-greenlake-terraform-provider” (#1893)
hpe-dev-incubator/hpe-dev-portal: content/blog/scale-kubernetes-clusters-using-hpe-greenlake-terraform-provider.md

---
title: Upgrade Kubernetes Clusters using HPE GreenLake Terraform Provider
date: 2023-05-12T04:07:48.888Z
author: Tanushi Agarwal
authorimage: /img/photoforblog.jpg
disable: false

> ---

Reviewer: Please add Tags here.
At the bottom of the CMS editor, you will see a section to add tags, "One at a time".
So, please add the tags you listed in the TAGS section at the end of your post, Make sure to add one tag at a time - you will have to click "Add Tags+" to add one tag. Repeat the same for each tag.
 And then remove the section "Tags" and its contents at the end of post.

IaC or Infrastructure as code is an practice of automating the process of managing and provisioning infrastructure through code instead of manual processes. It gives organizations tools to create, manage, and destroy compute resources by statically defining and declaring these resources in codeblocks. It helps to increase operational agility, simplify management, reduce errors, and save cost.

> > I﻿n this post, I will explore options to declare and upgrade Kubernetes clusters on HPE GreenLake using the HPE GreenLake Terraform Provider.

Reviewer: I would suggest to remove the character ">" so the sentence is aligned with the introductory paragraph above.

A﻿s the part of upgrade, the following 3 scenarios are supported --  

> 1. Scaling of a cluster's worker nodes. Please refer blog post [Scale Kubernetes Clusters using HPE GreenLake Terraform Provider](https://developer.hpe.com/blog/scale-kubernetes-cluster-using-hpe-greenlake-terraform-provider/) to check out scaling options available for woker nodes. 

Reviewer 2: ```suggestion
1. Scaling of a cluster's worker nodes. Please refer to the blog post [Scale Kubernetes Clusters using HPE GreenLake Terraform Provider](https://developer.hpe.com/blog/scale-kubernetes-cluster-using-hpe-greenlake-terraform-provider/) to check out scaling options available for worker nodes. 
```

> 1. Scaling of a cluster's worker nodes. Please refer blog post [Scale Kubernetes Clusters using HPE GreenLake Terraform Provider](https://developer.hpe.com/blog/scale-kubernetes-cluster-using-hpe-greenlake-terraform-provider/) to check out scaling options available for woker nodes. 

Reviewer 2: ```suggestion
1. Scaling of a cluster's worker nodes. Please refer to the blog post [Scale Kubernetes Clusters using HPE GreenLake Terraform Provider](https://developer.hpe.com/blog/scale-kubernetes-cluster-using-hpe-greenlake-terraform-provider/) to check out scaling options available for worker nodes. 
```

> 2. Upgrade the kubernetes version of the cluster. Covered in this blog.

Reviewer 2: ```suggestion
2. Upgrade the Kubernetes version of the cluster. This step is covered in this blog.
```

> 3. Upgrade the OS version of the worker nodes inside a cluster. Covered in this blog.

Reviewer 2: ```suggestion
3. Upgrade the OS version of the worker nodes inside a cluster. This step is covered in this blog.
```

# Prerequisite

> Before starting this tutorial, it is recommended that you read the blog post  [Kubernetes Cluster as Code - Part 1](https://developer.hpe.com/blog/kubernetes-clusters-as-code-part1/), which includes steps for creating a Kubernetes cluster. This post expands upon that scenario by examining how to upgrade a cluster's kubernetes version & OS version of its woker nodes. 

Reviewer 2: ```suggestion
Before starting this tutorial, read the blog post  [Kubernetes Cluster as Code - Part 1](https://developer.hpe.com/blog/kubernetes-clusters-as-code-part1/), which includes steps for creating a Kubernetes cluster. This post expands upon that scenario by examining how to upgrade a cluster's Kubernetes version & OS version of its worker nodes. 
```

# Verify existing Kubernetes cluster

After the cluster is created following the instructions found in the  [Kubernetes Cluster as Code - Part 1 blog post](https://developer.hpe.com/blog/kubernetes-clusters-as-code-part1/), launch HPE GreenLake Central console and verify that the cluster is present under the appropriate tenant.

You should see the  **tf-test**  cluster present under  **Dashboard -> Manage your Private Cloud -> Containers**.

![](https://developer.hpe.com/img/cluster_list.jpg)

Below is the reference Terraform configuration file for the existing cluster.

```hcl
terraform {
  required_providers {
    hpegl = {
      source  = "hpe/hpegl"
      version = ">= 0.2.2"
    }
  }
}
 
provider hpegl {
  caas {
  }
}
 
variable "HPEGL_SPACE" {
  type = string
}
 
data "hpegl_caas_site" "blr" {
  name = "BLR"
  space_id = var.HPEGL_SPACE
 }
 
data "hpegl_caas_cluster_blueprint" "bp" {
  name = "demo"
  site_id = data.hpegl_caas_site.blr.id
}
 
resource hpegl_caas_cluster test {
  name         = "tf-test"
  blueprint_id = data.hpegl_caas_cluster_blueprint.bp.id
  site_id      = data.hpegl_caas_site.blr.id
  space_id     = var.HPEGL_SPACE
}
```

> # Kubernetes version upgrade

Reviewer 2: Suggest this heading being in active voice: "Upgrade the Kubernetes version"?

> For Kubernetes version upgrade, we need to specify the new version of kubernetes that is available for upgrade in the resources block. 

Reviewer 2: ```suggestion
For the Kubernetes version upgrade, we need to specify the new version of Kubernetes that is available for upgrade in the resources block. 
```

Reviewer: Suggest to use "you need to specify,..." instead of "we need to specify...".

1. **kubernetes_version**: *Use the Kubernetes version that pops up on the cluster details page in the UI.*

   ![](/img/blog1.jpg)

>  Below is the reference Terraform configuration for updating the cluster's kubernetes version.

Reviewer 2: ```suggestion
 Below is the reference Terraform configuration for updating the cluster's Kubernetes version.
```

```hcl
terraform {
 required_providers {
   hpegl = {
     source  = "hpe/hpegl"
     version = ">= 0.2.2"
   }
 }
}

provider hpegl {
 caas {
 }
}

variable "HPEGL_SPACE" {
 type = string
}

data "hpegl_caas_site" "blr" {
 name = "BLR"
 space_id = var.HPEGL_SPACE
}

data "hpegl_caas_cluster_blueprint" "bp" {
 name = "demo"
 site_id = data.hpegl_caas_site.blr.id
}

resource hpegl_caas_cluster test {
 name         = "tf-test"
 blueprint_id = data.hpegl_caas_cluster_blueprint.bp.id
 site_id      = data.hpegl_caas_site.blr.id
 space_id     = var.HPEGL_SPACE
 kubernetes_version = "1.23.13-hpe1"
}
```

> # Ready to Terraform plan

Reviewer 2: Suggest using active voice: "Run Terraform plan"

Reviewer: Suggest to make it a level 2 header.

```shellsession
$ terraform plan

hpegl_caas_cluster.test: Refreshing state... [id=a32fabb9-7c19-42d1-9a38-ebf122810c0a]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # hpegl_caas_cluster.test will be updated in-place
  ~ resource "hpegl_caas_cluster" "test" {
        id                    = "a32fabb9-7c19-42d1-9a38-ebf122810c0a"
      ~ kubernetes_version    = "1.22.9-hpe1" -> "1.23.13-hpe1"
        name                  = "tf-test"
        # (17 unchanged attributes hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.

──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

Note: You didn't use the -out option to save this plan, so Terraform can't guarantee to take exactly these actions if you run "terraform apply" now.
```

> # Ready to Terraform apply

Reviewer 2: Suggest: "Run Terraform apply"

Reviewer: Suggest to make it a level 2 header.

hpegl_caas_cluster.test: Still modifying... [id=a32fabb9-7c19-42d1-9a38-ebf122810c0a, 19m10s elapsed]
hpegl_caas_cluster.test: Still modifying... [id=a32fabb9-7c19-42d1-9a38-ebf122810c0a, 21m10s elapsed]
hpegl_caas_cluster.test: Still modifying... [id=a32fabb9-7c19-42d1-9a38-ebf122810c0a, 23m10s elapsed]
hpegl_caas_cluster.test: Still modifying... [id=a32fabb9-7c19-42d1-9a38-ebf122810c0a, 25m10s elapsed]
hpegl_caas_cluster.test: Still modifying... [id=a32fabb9-7c19-42d1-9a38-ebf122810c0a, 27m10s elapsed]
hpegl_caas_cluster.test: Still modifying... [id=a32fabb9-7c19-42d1-9a38-ebf122810c0a, 29m10s elapsed]
hpegl_caas_cluster.test: Still modifying... [id=a32fabb9-7c19-42d1-9a38-ebf122810c0a, 31m10s elapsed]
hpegl_caas_cluster.test: Still modifying... [id=a32fabb9-7c19-42d1-9a38-ebf122810c0a, 33m10s elapsed]
hpegl_caas_cluster.test: Still modifying... [id=a32fabb9-7c19-42d1-9a38-ebf122810c0a, 35m10s elapsed]
hpegl_caas_cluster.test: Modifications complete after 35m18s [id=a32fabb9-7c19-42d1-9a38-ebf122810c0a]

Apply complete! Resources: 0 added, 1 changed, 0 destroyed.
```

# OS Version upgrade

> For OS version upgrade, we need to specify the new version of OS Image that is available for upgrade & the name of the OS Image in the worker node block.

Reviewer 2: ```suggestion
For an OS version upgrade, we need to specify the new version of the OS image that is available for upgrade & the name of the OS image in the worker node block.
```

Reviewer: Suggest to use "you need to..." rather than "we need to ..."

> **Note**: The OS version is specific to a worker node pool. All the nodes in the worker node pool will be updated to the same OS version. However, we can have different node pools supporting different OS versions. 

Reviewer: suggest "you can have different node pools" rather than "we can have different node pools...".

The worker node block consist of following fields - 

> 1. **name**: *The name of the worker nodepool. If you wish to update default or existing worker node pool, enter the name of that node pool or you can create a new node pool by specifying a new name.*

Reviewer 2: ```suggestion
1. **name**: *The name of the worker node pool. If you wish to update the default or existing worker node pool, enter the name of that node pool. To create a node pool, specify a new name.*
```

> 2. **machine_blueprint_id**: *Fill in the ID for the machine blueprint that is already present in HPE GreenLake Central for your tenant. Use the machine blueprint data source to retrieve the machine blueprint ID.*

Reviewer: for variables such as "machine_blueprint_id" you will need to escape the special "underscore" character.
The way to escape a special character is to add a backslash (\\) before it, for example: I do not want machine\\_blueprint\\_id.

3. **count**: *Add the number of nodes to be present as part of this node pool. We can scale up & down by updating the count value here.*
4. **os_image**: *The name of the OS image. Use the machine blueprint data source to retrieve the name of OS image.*
5. **os_version**: *The version to be upgraded to. We can get this version in the UI. Whenever there is a new version available, UI will pop up a banner mentioning the new version.*

> W﻿e can find the new version by clicking on the **Actions->Update OS** button on the right side of cluster's details page. 

Reviewer 2: ```suggestion
W﻿e can find the new version by clicking on the **Actions->Update OS** button on the right side of cluster details page. 
```

> ![](/img/blog2.jpg)

Reviewer: This image is not rendered in the preview. Please make sure the blog2.jpg is correctly uploaded into the CMS. You may want to try to upload it again in the CMS editor and select it as image.

![](/img/blog4.jpg)

>  Below is the reference Terraform configuration for updating the worker nodepool's OS version.

Reviewer 2: ```suggestion
 Below is the reference Terraform configuration for updating the OS version of the worker node pool.
```

```hcl
terraform {
 required_providers {
   hpegl = {
     source  = "hpe/hpegl"
     version = ">= 0.2.2"
   }
 }
}

provider hpegl {
 caas {
 }
}

variable "HPEGL_SPACE" {
 type = string
}

data "hpegl_caas_site" "blr" {
 name = "BLR"
 space_id = var.HPEGL_SPACE
}

data "hpegl_caas_cluster_blueprint" "bp" {
 name = "demo"
 site_id = data.hpegl_caas_site.blr.id
}

data "hpegl_caas_machine_blueprint" "standard_worker" {
 name = "standard-worker"
 site_id = "data.hpegl_caas_site.blr.id"
}

resource hpegl_caas_cluster test {
 name         = "tf-test"
 blueprint_id = data.hpegl_caas_cluster_blueprint.bp.id
 site_id      = data.hpegl_caas_site.blr.id
 space_id     = var.HPEGL_SPACE
 
worker_nodes {
   name = "test-node-pool"
   machine_blueprint_id = data.hpegl_caas_machine_blueprint.standard_worker.id
   count = "1"
   os_image = "sles-custom"
   os_version = "15.3"
 }
}
```

> # Ready to Terraform plan

Reviewer 2: Suggest: Run Terraform plan

Reviewer: suggest to make it a level2 header.

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # hpegl_caas_cluster.test will be updated in-place
  ~ resource "hpegl_caas_cluster" "test" {
        id                    = "a32fabb9-7c19-42d1-9a38-ebf122810c0a"
        name                  = "tf-test"
        # (18 unchanged attributes hidden)

      + worker_nodes {
          + count                = 1
          + machine_blueprint_id = "0ac21c99-2fdb-491d-a590-a5016690b80b"
          + name                 = "test-node-pool"
          + os_image             = "sles-custom"
          + os_version           = "15.3"
        }
    }

Plan: 0 to add, 1 to change, 0 to destroy.

──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

Note: You didn't use the -out option to save this plan, so Terraform can't guarantee to take exactly these actions if you run "terraform apply" now.
```

> # Ready to Terraform apply

Reviewer 2: Suggest: Run Terraform apply

Reviewer: suggest to make it a level 2 header.

hpegl_caas_cluster.test: Still modifying... [id=a32fabb9-7c19-42d1-9a38-ebf122810c0a, 25m10s elapsed]
hpegl_caas_cluster.test: Still modifying... [id=a32fabb9-7c19-42d1-9a38-ebf122810c0a, 27m10s elapsed]
hpegl_caas_cluster.test: Still modifying... [id=a32fabb9-7c19-42d1-9a38-ebf122810c0a, 29m10s elapsed]
hpegl_caas_cluster.test: Still modifying... [id=a32fabb9-7c19-42d1-9a38-ebf122810c0a, 31m10s elapsed]
hpegl_caas_cluster.test: Still modifying... [id=a32fabb9-7c19-42d1-9a38-ebf122810c0a, 33m10s elapsed]
hpegl_caas_cluster.test: Still modifying... [id=a32fabb9-7c19-42d1-9a38-ebf122810c0a, 35m10s elapsed]
hpegl_caas_cluster.test: Still modifying... [id=a32fabb9-7c19-42d1-9a38-ebf122810c0a, 37m10s elapsed]
hpegl_caas_cluster.test: Still modifying... [id=a32fabb9-7c19-42d1-9a38-ebf122810c0a, 39m10s elapsed]
hpegl_caas_cluster.test: Still modifying... [id=a32fabb9-7c19-42d1-9a38-ebf122810c0a, 41m10s elapsed]
hpegl_caas_cluster.test: Still modifying... [id=a32fabb9-7c19-42d1-9a38-ebf122810c0a, 43m10s elapsed]
hpegl_caas_cluster.test: Modifications complete after 43m18s [id=a32fabb9-7c19-42d1-9a38-ebf122810c0a]

Apply complete! Resources: 0 added, 1 changed, 0 destroyed.
```

> **Note**: If you wish to only scale up or down the worker node pool & not update the OS version, add only the 3 fields i.e. **name**, **machine_blueprint_id** & **count** in the worker node block. 

Reviewer: make sure to escape special character "underscore" for the variable machine_blueprint_id

# Summary

In this blog, I covered how to update Kubernetes clusters with Terraform provider for HPE GreenLake. I showed you how to update the Kubernetes version of the cluster & how to upgrade the OS version of worker node pool. Also I discussed how update & upgrade can be performed while scaling a cluster. 

I hope you found this information interesting and useful while considering the upgrade of Kubernetes cluster with HPE GreenLake Terraform provider. Use the following links to understand more about Terraform and HPE GreenLake Terraform Provider.

* [Learn more about Terraform](https://www.terraform.io/)
* [Learn more about HPE GreenLake](https://www.hpe.com/us/en/greenlake.html)
* [Learn more about the HPE GreenLake Terraform provider](https://registry.terraform.io/providers/HPE/hpegl)

Don’t forget, you can always find other tutorials and articles on HPE GreenLake on the  [HPE Developer blog](https://developer.hpe.com/blog/tag/hpe-greenlake).

> ## Tags

Reviewer: please add Tags using the CMS editor (just after the BODY section in the editor). Then delete the "Tags" section.

You should see the  **tf-test**  cluster present under  **Dashboard -> Manage your Private Cloud -> Containers**.

![](https://developer.hpe.com/img/cluster_list.jpg)

Below is the reference Terraform configuration file for the existing cluster.

```hcl
terraform {
  required_providers {
    hpegl = {
      source  = "hpe/hpegl"
      version = ">= 0.2.2"
    }
  }
}
 
provider hpegl {
  caas {
  }
}
 
variable "HPEGL_SPACE" {
  type = string
}
 
data "hpegl_caas_site" "blr" {
  name = "BLR"
  space_id = var.HPEGL_SPACE
 }
 
data "hpegl_caas_cluster_blueprint" "bp" {
  name = "demo"
  site_id = data.hpegl_caas_site.blr.id
}
 
resource hpegl_caas_cluster test {
  name         = "tf-test"
  blueprint_id = data.hpegl_caas_cluster_blueprint.bp.id
  site_id      = data.hpegl_caas_site.blr.id
  space_id     = var.HPEGL_SPACE
}
```

# Upgrade the Kubernetes version

For the Kubernetes version upgrade, you need to specify the new version of Kubernetes that is available for upgrade in the resources block. 

> 1. **kubernetes_version**: *Use the Kubernetes version that pops up on the cluster details page in the UI.*

Reviewer 3: I would use a regular font here... not italics.

> 1. **kubernetes_version**: *Use the Kubernetes version that pops up on the cluster details page in the UI.*

Reviewer 4: Is it possible to get that information programmatically? 
if yes, it would be interesting to mention it. Even if for this blog on terraform, the easy way is from the UI.

Author: No its not possible to get it programmatically.

hpegl_caas_cluster.test: Still modifying... [id=a32fabb9-7c19-42d1-9a38-ebf122810c0a, 27m10s elapsed]
hpegl_caas_cluster.test: Still modifying... [id=a32fabb9-7c19-42d1-9a38-ebf122810c0a, 29m10s elapsed]
hpegl_caas_cluster.test: Still modifying... [id=a32fabb9-7c19-42d1-9a38-ebf122810c0a, 31m10s elapsed]
hpegl_caas_cluster.test: Still modifying... [id=a32fabb9-7c19-42d1-9a38-ebf122810c0a, 33m10s elapsed]
hpegl_caas_cluster.test: Still modifying... [id=a32fabb9-7c19-42d1-9a38-ebf122810c0a, 35m10s elapsed]
hpegl_caas_cluster.test: Modifications complete after 35m18s [id=a32fabb9-7c19-42d1-9a38-ebf122810c0a]

Apply complete! Resources: 0 added, 1 changed, 0 destroyed.
```

# U﻿pgrade the OS Version

For an OS version upgrade, you need to specify the new version of the OS image that is available for upgrade & the name of the OS image in the worker node block.

**Note**: The OS version is specific to a worker node pool. All the nodes in the worker node pool will be updated to the same OS version. However, you can have different node pools supporting different OS versions. 

The worker node block consists of the following fields -

> 1. **name**: *The name of the worker node pool. If you wish to update the default or existing worker node pool, enter the name of that node pool. To create a node pool, specify a new name.*

Reviewer 3: In all of the bullets below, I would recommend not using the italic font. It makes it a little difficult to read. Generally, one uses italics to emphasize something, usually one or two words in a sentence, or to highlight an important note.

2. **machine\_blueprint\_id**: *Fill in the ID for the machine blueprint that is already present in HPE GreenLake Central for your tenant. Use the machine blueprint data source to retrieve the machine blueprint ID.*
3. **count**: *Add the number of nodes to be present as part of this node pool. We can scale up & down by updating the count value here.*
4. **os_image**: *The name of the OS image. Use the machine blueprint data source to retrieve the name of OS image.*
5. **os_version**: *The version to be upgraded to. We can get this version in the UI. Whenever there is a new version available, UI will pop up a banner mentioning the new version.*

W﻿e can find the new version by clicking on the **Actions->Update OS** button on the right side of cluster details page. 

![](/img/blog3.jpg)

> 

Author: No there is no program way of finding out the new version available for upgrade. we can only find existing version information programmatically.

![](/img/blog4.jpg)

Below is the reference Terraform configuration for updating the OS version of the worker node pool.

```hcl
terraform {
 required_providers {
   hpegl = {
     source  = "hpe/hpegl"
     version = ">= 0.2.2"
   }
 }
}

provider hpegl {
 caas {
 }
}

variable "HPEGL_SPACE" {
 type = string
}

data "hpegl_caas_site" "blr" {
 name = "BLR"
 space_id = var.HPEGL_SPACE
}

data "hpegl_caas_cluster_blueprint" "bp" {
 name = "demo"
 site_id = data.hpegl_caas_site.blr.id
}

data "hpegl_caas_machine_blueprint" "standard_worker" {
 name = "standard-worker"
 site_id = "data.hpegl_caas_site.blr.id"
}

resource hpegl_caas_cluster test {
 name         = "tf-test"
 blueprint_id = data.hpegl_caas_cluster_blueprint.bp.id
 site_id      = data.hpegl_caas_site.blr.id
 space_id     = var.HPEGL_SPACE
 
worker_nodes {
   name = "test-node-pool"
   machine_blueprint_id = data.hpegl_caas_machine_blueprint.standard_worker.id
   count = "1"
   os_image = "sles-custom"
   os_version = "15.3"
 }
}
```

## Run Terraform plan

> Again, to preview the changes, do a dry run by running **terraform plan**

Reviewer 3: Add a period at the end of this sentence.

hpegl_caas_cluster.test: Still modifying... [id=a32fabb9-7c19-42d1-9a38-ebf122810c0a, 43m10s elapsed]
hpegl_caas_cluster.test: Modifications complete after 43m18s [id=a32fabb9-7c19-42d1-9a38-ebf122810c0a]

Apply complete! Resources: 0 added, 1 changed, 0 destroyed.
```

**Note**: If you wish to only scale up or down the worker node pool & not update the OS version, add only the 3 fields i.e. **name**, **machine\_blueprint\_id** & **count** in the worker node block. 

# Summary

In this blog, I covered how to update Kubernetes clusters with Terraform provider for HPE GreenLake. I showed you how to update the Kubernetes version of the cluster & how to upgrade the OS version of worker node pool. Also I discussed how update & upgrade can be performed while scaling a cluster. 

I hope you found this information interesting and useful while considering the upgrade of Kubernetes cluster with HPE GreenLake Terraform provider. Use the following links to understand more about Terraform and HPE GreenLake Terraform Provider.

* [Learn more about Terraform](https://www.terraform.io/)
* [Learn more about HPE GreenLake](https://www.hpe.com/us/en/greenlake.html)

> * [Learn more about the HPE GreenLake Terraform provider](https://registry.terraform.io/providers/HPE/hpegl)

Reviewer 4: suggestion (minor) to link the doc page of the terraform provider : 
https://registry.terraform.io/providers/HPE/hpegl/latest/docs
